include/odp/api/spec/dma_types.h

```diff
@@ -0,0 +1,547 @@
+/* Copyright (c) 2021, Nokia
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+/**
+ * @file
+ *
+ * ODP DMA
+ */
+
+#ifndef ODP_API_SPEC_DMA_TYPES_H_
+#define ODP_API_SPEC_DMA_TYPES_H_
+#include <odp/visibility_begin.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <odp/api/std_types.h>
+
+/** @defgroup odp_dma ODP DMA
+ *  DMA offload
+ *  @{
+ */
+
+/**
+ * @typedef odp_dma_t
+ * DMA session
+ */
+
+/**
+ * @typedef odp_dma_transfer_id_t
+ * DMA transfer identifier
+ */
+
+/**
+ * @typedef odp_dma_compl_t
+ * DMA completion event
+ */
+
+/**
+ * @def ODP_DMA_INVALID
+ * Invalid DMA session
+ */
+
+/**
+ * @def ODP_DMA_TRANSFER_ID_INVALID
+ * Invalid DMA transfer identifier
+ */
+
+/**
+ * @def ODP_DMA_COMPL_INVALID
+ * Invalid DMA completion event
+ */
+
+/**
+ * @def ODP_DMA_NAME_LEN
+ * Maximum DMA name length in chars including null char
+ */
+
+/**
+ * DMA completion event pool capabilities
+ *
+ * Pool statistics are not supported with DMA completion event pools.
+ */
+typedef struct odp_dma_pool_capability_t {
+	/** Maximum number of DMA completion event pools */
+	uint32_t max_pools;
+
+	/** Maximum number of DMA completion events in a pool */
+	uint32_t max_num;
+
+	/** Minimum size of thread local cache */
+	uint32_t min_cache_size;
+
+	/** Maximum size of thread local cache */
+	uint32_t max_cache_size;
+
+} odp_dma_pool_capability_t;
+
+/**
+ * DMA completion event pool parameters
+ */
+typedef struct odp_dma_pool_param_t {
+	/** Number of DMA completion events in the pool
+	 *
+	 *  Maximum value is defined by 'max_num' pool capability */
+	uint32_t num;
+
+	/** Maximum number of events cached locally per thread
+	 *
+	 *  See odp_pool_param_t::cache_size documentation for details. Valid values range from
+	 *  'min_cache_size' to 'max_cache_size' capability. The default value is implementation
+	 *  specific and set by odp_dma_pool_param_init().
+	 */
+	uint32_t cache_size;
+
+} odp_dma_pool_param_t;
+
+/* Includes pool_types.h, which depends on odp_dma_pool_param_t. */
+#include <odp/api/queue.h>
+
+/**
+ * DMA transfer direction
+ *
+ * Transfer direction defines source and destination memory type of DMA transfers. API specification
+ * defines only one option (#ODP_DMA_MAIN_TO_MAIN) for the transfer direction. It is used for
+ * transfers within the main memory. Some implementations may extend this enumeration with
+ * implementation specific directions and memory types (e.g. from main memory to a device, etc.).
+ */
+typedef uint32_t odp_dma_direction_t;
```

Comment:
Only main to main DMA transfer direction is supported by the spec.
Platforms are free to extend and implement Main to device, device to main and device to device.
Would it be better to have all 4 directions in main spec and platforms can chose to specify the supported directions in capabilities?
That way app would not have to be aware of platform to use extended implementation.

Reply:
I left other directions (non-main memory) intentionally not specified in this initial version of the DMA API. We can add those later when needed. When adding non-main memory support, we'd need to solve couple of questions like: how application gets non-main memory addresses that can be used with ODP DMA API in a portable way (through Linux, new ODP APIs, ...), are those addresses virtual or physical, etc